plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,372 @@
+  #servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	//log.Println("[outputs.whatap] Connect")
+	w.Session.Dest += 1
+	if w.Session.Dest >= len(w.Servers) {
+		w.Session.Dest = 0
+	}
+
+	addr := strings.SplitN(w.Servers[w.Session.Dest], "://", 2)
+	if len(addr) != 2 {
+		return fmt.Errorf("invalid address: %s", w.Servers[w.Session.Dest])
+	}
+	if addr[0] != "tcp" {
+		return fmt.Errorf("only tcp is supported: %s", w.Servers[w.Session.Dest])
+	}
+
+	t := w.Timeout * time.Millisecond
+	if client, err := net.DialTimeout(addr[0], addr[1], t); err != nil {
+		return err
+	} else {
+		w.Session.Client = client.(*net.TCPConn)
+		log.Println("I! [outputs.whatap] Connected tcp to ", addr)
+	}
+
+	return nil
+
+}
+
+func (w *Whatap) Close() error {
+	//log.Println("[outputs.whatap] Closed")
+	if w.Session.Client == nil {
+		return nil
+	}
+	err := w.Session.Client.Close()
+	w.Session.Client = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Configuration for WhaTap"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	//log.Println("[outputs.whatap] Write len=", len(metrics))
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.Session.Client == nil {
+		if err := w.Connect(); err != nil {
+			w.Close()
+			return err
+		}
+	}
+	for _, m := range metrics {
+		p := NewTagCountPack()
+		p.Pcode = w.Pcode
+		p.Oid = w.Oid
+		p.Category = fmt.Sprintf("%s%s", "Telegraf.", m.Name())
+		for k, v := range m.Fields() {
+			p.Put(k, v)
+		}
+		for k, v := range m.Tags() {
+			p.PutTag(k, v)
+		}
+		// Add Oname
+		p.PutTag("oname", w.Oname)
+		p.Time = m.Time().UnixNano() / int64(time.Millisecond)
+
+		dout := NewDataOutputX()
+		// pack type
+		dout.WriteShort(p.GetPackType())
+		p.Write(dout)
+
+		if err := w.send(0, dout.ToByteArray()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (w *Whatap) send(code byte, b []byte) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("Recover Whatap: %s", r)
+		}
+	}()
+
+	out := NewDataOutputX()
+	out.WriteByte1(NETSRC_AGENT_ONEWAY)
+	// ver
+	out.WriteByte1(code)
+	out.WriteLong(w.Pcode)
+	out.WriteLong(Hash64Str(w.License))
+	out.WriteIntBytes(b)
+	// pack data
+	sendbuf := out.ToByteArray()
+
+	nbyteleft := len(sendbuf)
+	pos := 0
+	for 0 < nbyteleft {
+		nbytethistime := 0
+		// Set Deadline
+		err = w.Session.Client.SetWriteDeadline(time.Now().Add(5 * time.Millisecond))
+		if err != nil {
+			log.Println("W! [outputs.whatap] cannot set tcp write deadline:", err)
+		}
+		nbytethistime, err = w.Session.Client.Write(sendbuf[pos : pos+nbyteleft])
+		if err != nil {
+			w.Close()
+			return err
+		}
+		nbyteleft -= nbytethistime
+		pos += nbytethistime
+	}
+	return err
+
+}
+
+var table = [...]int64{
+	0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419,
```

Comment:
What is this?

Reply:
TCP 서버와 공용으로 사용하는 hash table 입니다. 캐싱또는 ID처럼 사용하는 값으로 서버와 동일한 로직으로 되어 있어야 합니다.

Reply:
So this is a random value shared with the server implementation?  Nobody would be able to figure this out, so if we are to include it then at a minimum it would need to be well documented in a code comment as well as in https://docs.whatap.io/kr/agent_telegraf/api.html

Reply:
I added a comment to the code for hash table.
Added a description of the hash format and a code description to the api documentation.
```

/**
 * The following table array is a hash lookup table.
 * Generate int32, int64 hash of string or byte array based on the table.
 *
 * The generated hash must be the same as the WhaTap server side.
 * (The same hash lookup table and function are on the WhaTap server.)
 * This is because the WhaTap server caches or indexes data
 * with that hash value.
 *
 * The HashBytes, HashStr function converts a string or byte array
 * to an int32 hash value.
 *
 * The Hash64Bytes, Hash64Str function converts a string or byte array
 * to an int64 hash value.
 */
var table = [...]int64{

```